packages/@react-stately/grid/src/GridCollection.ts

```diff
@@ -23,92 +23,106 @@ export class GridCollection<T> implements IGridCollection<T> {
 
   constructor(opts?: GridCollectionOptions<T>) {
     this.keyMap = new Map();
     this.columnCount = opts?.columnCount;
     this.rows = [];
+    this.sections = [];
+
+    let visit = (node: GridNode<T>, i?: number) => {
+      // Reset row counter if entering a new section so we have a index reflecting the position of the
+      // row within the section
+      if (node.type === 'section') {
+        rowCounter = 0;
+      }
 
-    let visit = (node: GridNode<T>) => {
-      // If the node is the same object as the previous node for the same key,
-      // we can skip this node and its children. We always visit columns though,
-      // because we depend on order to build the columns array.
-      let prevNode = this.keyMap.get(node.key);
       if (opts.visitNode) {
         node = opts.visitNode(node);
       }
 
-      this.keyMap.set(node.key, node);
-
-      let childKeys = new Set();
-      let last: GridNode<T>;
-      for (let child of node.childNodes) {
-        if (child.type === 'cell' && child.parentKey == null) {
-          // if child is a cell parent key isn't already established by the collection, match child node to parent row
-          child.parentKey = node.key;
-        }
-        childKeys.add(child.key);
+      if (node.type === 'headerrow' || node.type === 'item' || node.type === 'section') {
+        // TODO: perhaps only add some of these default props for node.type === 'item'. Previously this was done fo all items provided
+        // to the GridCollection...
+        let defaultProps = {
+          level: 0,
+          key: 'row-' + i,
+          type: 'row',
+          value: undefined,
+          hasChildNodes: true,
+          childNodes: [...node.childNodes],
+          rendered: undefined,
+          textValue: undefined,
+          colspan: node.type === 'section' || node.props?.isSectionHeader ? this.columnCount : 1
+        };
+
+        let newProps = {
+          index: node.type === 'section' ? sectionCounter++ : rowCounter++,
+          rowIndex: node.type !== 'section' ? rowIndex : undefined
+        };
+
+        // Use Object.assign instead of spread to preseve object reference for keyMap. Also ensures retrieving nodes
+        // via .childNodes returns the same object as well
+        Object.assign(node, defaultProps, {...node}, newProps);
+      }
 
+      if (node.type !== 'cell') {
         if (last) {
-          last.nextKey = child.key;
-          child.prevKey = last.key;
-        } else {
-          child.prevKey = null;
+          last.nextKey = node.key;
+          node.prevKey = last.key;
         }
-
-        visit(child);
-        last = child;
+        last = node;
       }
 
-      if (last) {
-        last.nextKey = null;
-      }
+      this.keyMap.set(node.key, node);
 
-      // Remove deleted nodes and their children from the key map
-      if (prevNode) {
-        for (let child of prevNode.childNodes) {
-          if (!childKeys.has(child.key)) {
-            remove(child);
```

Comment:
Not sure when this old code comes into play, ran all the tests on main and never seems to fire. None of the table stories call it either